src/main/scala/workflow/AutoCacheRule.scala

```diff
@@ -0,0 +1,355 @@
+            // Construct a
+            val sample = transformer.transformRDD(sampledInputs).cache()
+            sample.count()
+            val duration = System.nanoTime() - start
+
+            // Profile sample memory
+            val memSize = sample.context.getRDDStorageInfo.filter(_.id == sample.id).map(_.memSize).head
+
+            // If this sample was computed using the final and largest scale, add it to the registers
+            if ((scaleIndex == (sortedScales.length - 1)) && (trial == numTrials)) {
+              registers(i) = RDDOutput(sample)
+            } else {
+              sample.unpersist()
+            }
+
+            SampleProfile(scaledNumPerPartition.values.sum, Profile(duration, memSize))
+          }
+
+          profiles(i) = generalizeProfiles(totalCount, sampleProfiles)
+        }
+
+        case node: Instruction => {
+          val deps = node.getDependencies.map(registers)
+          registers(i) = node.execute(deps)
+        }
+      }
+    }
+
+    profiles.toMap
+  }
+
+
+  /**
+   * Estimates the total runtime of a pipeline given the cached set of instructions
+   */
+  def estimateCachedRunTime(
+    instructions: Seq[Instruction],
+    cached: Set[Int],
+    profiles: Map[Int, Profile]
+  ): Double = {
+    val nodeWeights = getNodeWeights(instructions)
+    val runs = getRuns(instructions, cached, nodeWeights)
+    val localWork = instructions.indices.map(i => profiles.getOrElse(i, Profile(0, 0)).ns.toDouble).toArray
+
+    instructions.indices.map(i => {
+      val executions = if (cached(i)) 1 else runs(i)
+      localWork(i) * executions
+    }).sum
+  }
+
+  /**
+   * Given a seq of instructions and a set of indices to cache - return an instruction seq with the indices cached.
+   */
+  def makeCachedPipeline(pipe: Seq[Instruction], cached: Set[Int]): Seq[Instruction] = {
+    // Find the indexes of the new caching nodes. We only cache instructions that produce RDDs
+    val dataOutputtingInstructions = pipe.zipWithIndex.filter {
+      case (TransformerApplyNode(_, _), _) => true
+      case (SourceNode(_), _) => true
+      case _ => false
+    }.map(_._2).toSet
+
+    val toCache = cached.intersect(dataOutputtingInstructions)
+
+    pipe.indices.foldLeft (
+      (Seq[Instruction](), pipe.indices.zipWithIndex.toMap + (Pipeline.SOURCE -> Pipeline.SOURCE))
+    ) {
+      case ((newPipe, oldToNewIndexMap), i) if toCache.contains(i) =>
+        (newPipe ++ Seq(
+          pipe(i).mapDependencies(oldToNewIndexMap),
+          new Cacher,
+          TransformerApplyNode(oldToNewIndexMap(i) + 1, Seq(oldToNewIndexMap(i)))
+        ),
+          oldToNewIndexMap.map {
+            case (key, value) => if (key >= i) (key, value + 2) else (key, value)
+          })
+
+      case ((newPipe, oldToNewIndexMap), i) =>
+        (newPipe :+ pipe(i).mapDependencies(oldToNewIndexMap), oldToNewIndexMap)
+    }._1
+  }
+
+  def naiveCache(instructions: Seq[Instruction]): Seq[Instruction] = {
```

Comment:
Can we call this something other than `naive` since that's not very descriptive anyway? Something that conveys that we're caching everything that is being reused and with an LRU replacement policy.